File: Data Structures & Algorithms/same-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        def same(first, second):
            if not first and not second:
                return True
            if not first or not second:
                return False
            if first.val != second.val:
                return False
            left = same(first.left, second.left)
            if not left:
                return False
            right = same(first.right, second.right)
            if not right:
                return False
            return True
        
        return same(p,q)

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {

private:
    int height(TreeNode* node){
        if (node==nullptr){
            return 0;
        }
        int left = height(node->left);
        if (left==-1){
            return -1;
        }
        int right = height(node->right);
        if (right==-1){
            return -1;
        }

        if (abs(right-left)>1){
            return -1;
        }

        return 1+max(left,right);

    }
public:
    bool isBalanced(TreeNode* root) {
        if (root==nullptr){
            return true;
        }

        int leftheight = height(root->left);
        if (leftheight==-1){
            return false;
        }
        int rightheight = height(root->right);
        if (rightheight==-1){
            return false;
        }
        if (abs(rightheight-leftheight)>1){
            return false;
        }

        return true;
        
    }
};
